File: doc/doxy/preprocmd.pl
# doxygen is annoying about certain .md things so we have to
# preprocess before doxygenating

use strict;
use warnings;

my ($infile, $outfile) = @ARGV;
my ($infh, $outfh);

my $contents = do {
        open( $infh, '<', $infile ) or die "Can't open $infile: $!";
        local $/ = undef;
        <$infh>;
    };
close $infh;

$contents =~ s/\(\.\//(/g;                       # ./ -based relative paths
$contents =~ s/##\s\+\[/* [/g;                   # links in headings
$contents =~ s/src=\"\.\/dot\//SRC="..\/dot\//g; # relative locations of images

my ($x, $y, $z);

# doxygen doesn't make heading anchors like it should
$contents =~  s/(^\#+\s+)(.*)/$y=$1;$z=$2;($x=lc($2))=~s! !-!g;$x=~s!,!!g;"$y $z\n\<a name=\"$x\" id=\"$x\"\>\<\/a\>"/gem;

# doxygen is specifically annoying about fenced code, which only
# occurs in README.MAIN.md
# - and C++ has to be converted to doxygen \code ...\endcode to get linking
$contents =~ s/\n( +)```bash(\n.*?)\n\1```/($x=$2)=~s!\n!\n    !g;$x/gse;
$contents =~ s/```c\+\+(\n.*?\n)```/\\code{.cpp}$1\\endcode/gs;

open( $outfh, '>', $outfile ) or die "Can't open $outfile: $!";
print $outfh $contents;


close $outfh;

print "md preprocessing done\n\n"

File: doc/doxy/preprocmd.py
# doxygen is annoying about certain .md things so we have to
# preprocess before doxygenating

import re
import sys


(infile, outfile) = sys.argv[1:3]

def anchor_heading(headtxt, anchortxt):
    return '{0}<a name="{1}" id="{1}"></a>\n\n'.format(
         headtxt,
         reduce (lambda intxt,(oldtxt,newtxt): intxt.replace(oldtxt,newtxt),
                  [anchortxt.lower(),
                   (', ',''), ('/',''), (' ', '-')
                   ]
                  ))

with open(infile,'r') as infh:
    contents = infh.read()
    
    contents = reduce(
        lambda intxt, (inre,replacement): re.sub(inre,replacement,intxt),
        [contents,
         
         # doxygen doesn't make heading anchors like it should
         (r'(?:(?<=\n)|^)(\#+\s+)(.*)\n',
          lambda m:anchor_heading(m.group(0), m.group(2))),

         (r'(?:(?<=\n)|^)(\S(.*))\n(?:(\=\=\=+|\-\-\-+)\n)',
          lambda m:anchor_heading(m.group(0), m.group(1))), 

     
         # doxygen is specifically annoying about fenced code, which
         # only occurs in README.md. 
         (r'\n( +)```bash(\n(.*\n)*?)\1```',
          lambda m: m.group(1) + m.group(2).replace('\n','\n    ')),


         # C++ has to be converted to doxygen \code ...\endcode to get
         # linking
         (r'```c\+\+(\n(.*\n)*?)```',
          r'\\code{.cpp}\1\\endcode')
         ]
        )

   # print contents


    with open(outfile,'w') as outfh:
        outfh.write(contents)

print "md preprocessing done\n\n"



